### ISSUE TITLE
Support for user defined callback functions

### ISSUE THREAD

**Issue by tkarna** [thetisproject/thetis#53]
In many cases users want to compute custom diagnostics of simulation state at run time. It would be useful to have a generic method for supporting this.

I think all the existing callbacks (volume and mass conservation etc) could then be ported to the same interface.

**Comment by tkarna** [comment]
The interface could be something like this.

User defines a callback object:

```
class TotalSaltCallback(thetis.somethingsomething):
    """Simple callback that computes an integral of salinity field"""
    def __init__(self):
        self.init_value = None

    def evaluate(solver_obj):
        """Defines the actual functional"""
        s = assemble(solver_obj.fields.salinity_3d*dx)
        if self.init_value is None:
            self.init_value = s
        rel_diff = (s - self.init_value)/self.init_value
        return (s, rel_diff)

    def info_str(self, t):
        """Used defined string that will be printed at run time, if registered with
        print_to_log=True."""
        return 'Total salt {:.4f}, rel. diff. {:e}'.format(t[0], t[1])

```

The callback can then be added to the solver:

```
solver_obj = ...
scomputer = TotalSaltCallback()

solver_obj.add_callback('totalsalt', scomputer,
                        interval='export', print_to_log=True, save_to_disk=True)
```

The interval argument defines when the callback should be evaluated. At least options `export` and `timestep`, for evaluating at every export or every time step, seem reasonable.

If `save_to_disk` is `False` no diagnostic output file will be created.

One issue is how to deal with the file output, that is, where to store the outputs and in which file format.

For instance we could store the data in hdf5 files:

```
outputdir/diagnostics/totalsalt.hdf5
```

where `outputdir` is the directory used for the exports. The file contains the time stamps and callback values. Alternatively we could store the output in ascii files (the files are presumably going to be small) but as we have hdf5 support I'd prefer that.

**Comment by wence-** [comment]
I would be inclined to have an interface like this:

```python

class DiagnosticCallback(object):
    __metaclass__ = ABCMeta

   def __init__(self):
       pass

   @abstractproperty
   def name(self):
       "The name of the diagnostic"
       pass

   @abstractmethod
   def __call__(self, solver_obj):
       """Evaluate the diagnostic value.
    
       :arg solver_obj: The solver object.
       """
       pass

   @abstractmethod
   def __str__(self, *args):
        """A string representation.

        :arg args: If provided, these will be the return value from :meth:`__call__`.
        """
        return "{} diagnostic".format(self.name)


class TotalSalt(DiagnosticCallback):
    name = "totalsalt"

    def __call__(self, solver_object):
       ...

    def __str__(self, t0, t1):
        return "Total Salt ...".format(t0, t1)

solver_obj.add_callback(TotalSalt(), ...)
```

**Comment by tkarna** [comment]
Yes, that's more pythonic. I wonder if the `name` property is necessary, one could just use `__class__.__name__` in the abstract class?

**Comment by wence-** [comment]
That's true:

```python
@property
def name(self):
   return type(self).__name__
```

**Comment by tkarna** [comment]
This is now drafted in #55. With hdf5 outputs the structure is a bit different from what's outlined above.

**Comment by tkarna** [comment]
This is now implemented in #55 and #66.
